### PR TITLE
feat: Athena Databases and Tables

### PR DESCRIPTION
**Pull request by richardhboyd**
WIP for Athena Resources via AwsCustomResources
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-athena/lib/datatypes.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+   */
+  ARRAY = 'ARRAYTODO',
+  /**
+   * < primitive_type, data_type >
+   */
+  MAP = 'MAPTODO',
+  /**
+   * < col_name : data_type [COMMENT col_comment] [, ...] >
+   */
+  STRUCT = 'STRUCTTODO'
```
Comment (richardhboyd) [inline packages/@aws-cdk/aws-athena/lib/datatypes.ts]: I need a better way to represent these data structures inside the schema.

File: packages/@aws-cdk/aws-athena/lib/table.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+export interface TableProps {
+  /**
+   * The database to add the Table to.
+   * @default none
+   */
+  readonly databaseName: string;
+
+  /**
+   * The name for the Table.
+   * @default none
```
Comment (richardhboyd) [inline packages/@aws-cdk/aws-athena/lib/table.ts]: Do I need '@default' for required properties?

File: packages/@aws-cdk/aws-athena/lib/table.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+   * some docstring
+   */
+  public readonly tableName: string;
+
+  constructor(scope: Construct, id: string, props: TableProps) {
+  super(scope, id);
+  this.tableName = props.tableName;
+
+  const s3Policy = new PolicyStatement();
+  s3Policy.addActions('s3:*');
```
Comment (richardhboyd) [inline packages/@aws-cdk/aws-athena/lib/table.ts]: Need to scope this down to the read Actions

File: packages/@aws-cdk/aws-athena/lib/table.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  const gluePolicy = new PolicyStatement();
+  gluePolicy.addActions('glue:GetDatabase', 'glue:GetTable');
+  athenaPolicy.addResources(`arn:aws:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:catalog`);
+
+  new AwsCustomResource(this, 'CreateAthenaTable', {
+    onCreate: {
+      service: 'Athena',
+      action: 'startQueryExecution',
+      parameters: {
+        QueryString: `CREATE EXTERNAL TABLE IF NOT EXISTS ${props.databaseName}.${props.tableName} (${schemaStringBuilder(props.schema)}) LOCATION 's3://${props.queryBucket.bucketName}'/`,
```
Comment (richardhboyd) [inline packages/@aws-cdk/aws-athena/lib/table.ts]: Add the Comments arguments to the query

File: packages/@aws-cdk/aws-athena/lib/table.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        ResultConfiguration: {
+        OutputLocation: `s3://${props.queryBucket.bucketName}/`
```
Comment (richardhboyd) [inline packages/@aws-cdk/aws-athena/lib/table.ts]: I need a better place for this. we really don't care about the output of this query but this field is required.

**Comment by netroy** [comment]
I really like that CDK is finally getting more Athena support. Thanks ❤️.
However, I'd like to point out a few things.

### 1. Deprecation of Athena's original metastore
Since Glue has been available across all regions, `CREATE DATABASE` & `CREATE TABLE` queries on Athena, actually create databases & tables on the Glue Data Catalog (instead of the metastore Athena originally launched with).
It might make sense to create Athena `Database` & `Table` classes as wrappers on top of the already existing classes in `@aws-cdk/aws-glue`, instead of making custom-resources for them.

[More info](https://docs.aws.amazon.com/athena/latest/ug/glue-faq.html)

### 2. Avro has special schema requirements
For Apache Avro, column definitions alone don't work right now in Athena. 
An Avro style schema also needs to be given, 
* either as `Table.StorageDescriptor.SerDeInfo.Parameters` if creating the Table on Glue
* or, `SERDEPROPERTIES` in the `CREATE EXTERNAL TABLE` query, if creating the Table over Athena, or Spectrum.

[More info](https://docs.aws.amazon.com/athena/latest/ug/avro.html)

**Comment by richardhboyd** [comment]
This is great feedback. I was just working on trying to understand how Glue Tables and Athena Tables were related.

On Nov 28, 2019, at 2:05 PM, करतोफ्फेलस्क्रिप्ट™ <notifications@github.com> wrote:

﻿

I really like that CDK is finally getting more Athena support. Thanks ❤️.
However, I'd like to point out a few things.

1. Deprecation of Athena's original metastore

Since Glue has been available across all regions, CREATE DATABASE & CREATE TABLE queries on Athena, actually create databases & tables on the Glue Data Catalog (instead of the metastore Athena originally launched with).
It might make sense to create Athena Database & Table classes as wrappers on top of the already existing classes in @aws-cdk/aws-glue, instead of making custom-resources for them.

More info<https://docs.aws.amazon.com/athena/latest/ug/glue-faq.html>

2. Avro has special schema requirements

For Apache Avro, column definitions alone don't work right now in Athena.
An Avro style schema also needs to be given,

  *   either as Table.StorageDescriptor.SerDeInfo.Parameters if creating the Table on Glue
  *   or, SERDEPROPERTIES in the CREATE EXTERNAL TABLE query, if creating the Table over Athena, or Spectrum.

More info<https://docs.aws.amazon.com/athena/latest/ug/avro.html>

**Comment by netroy** [comment]
@richardhboyd I created [this PR](https://github.com/aws/aws-cdk/pull/5246) earlier today. 
It might be relevant in this context.
Perhaps, I can have your feedback 🙂

File: packages/@aws-cdk/aws-athena/lib/table.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-athena/lib/table.ts]: No, it doesn't really make sense to have `@default` for required properties - feel free to remove these.

File: packages/@aws-cdk/aws-athena/lib/table.ts (hunk 3 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-athena/lib/table.ts]: Also, you can create a `PolicyStatement` in one expression:

```ts
  new AwsCustomResource(this, 'CreateAthenaTable', {
    // ...
    policyStatements: [
      new PolicyStatement({
        actions: [...],
        resources: [...],
      }),
      // ...
    ],
  });
```

File: packages/@aws-cdk/aws-athena/package.json [hunk 6]
```diff
@@ -68,14 +68,18 @@
     "pkglint": "1.18.0"
   },
   "dependencies": {
-    "@aws-cdk/core": "1.18.0"
```
Comment (skinny85) [inline packages/@aws-cdk/aws-athena/package.json]: Leave the `@aws-cdk/core` dependency (this module uses a bunch of things from it, like `Construct`).

Comment (richardhboyd) [reply packages/@aws-cdk/aws-athena/package.json]: It was just pushed down a line so the dependencies are in alphabetical order

File: packages/@aws-cdk/aws-athena/lib/table.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+      policyStatements: [s3Policy, athenaPolicy, gluePolicy]
+  });
+  }
+}
+
+function schemaStringBuilder(schema: {[key: string]: DataType}) {
+  const tempStrings: string[] = [];
+  tempStrings.concat(Object.keys(schema)
+    .map(key => `'${key}' '${schema[key]}'`));
+  return `(${tempStrings.join(", ")})`;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-athena/lib/table.ts]: This is a very surprising way to write this code... `concat` doesn't modify its receiver, it returns a new array, and also you're calling `concat` on something you know is empty.

Did you mean something like this?

```ts
function schemaStringBuilder(schema: { [key: string]: DataType }): string {
  return '(' +
    Object.keys(schema)
      .map(key => `'${key}' '${schema[key]}'`))
      .join(', ', '(', ')')})`
    + ')';
}
```

Comment (richardhboyd) [reply packages/@aws-cdk/aws-athena/lib/table.ts]: ohhh, that's good to know. I'll make that change.

**Comment by netroy** [comment]
@richardhboyd can I somehow help move this forward?
If there is an RFC, I could maybe send some PRs.

Cheers

File: packages/@aws-cdk/aws-athena/lib/database.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  s3Policy.addResources(props.locationBucket.arnForObjects('*'));
+  s3Policy.addResources(props.locationBucket.bucketArn);
+
+  const athenaPolicy = new PolicyStatement();
+  athenaPolicy.addActions('athena:StartQueryExecution');
+  athenaPolicy.addAllResources();
+
+  const gluePolicy = new PolicyStatement();
+  gluePolicy.addActions('glue:GetDatabase', 'glue:GetTable');
+  athenaPolicy.addResources(`arn:aws:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:catalog`);
```
Comment (ayenter) [inline packages/@aws-cdk/aws-athena/lib/database.ts:77]: I think this should be `gluePolicy.addResources(...`

**Comment by skinny85** [comment]
I'm closing this one due to inactivity, please comment if you want it re-opened.

**Pull request closed without merging**